fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,351 @@
+                .iter()
+                .map(|source| self.update_source_continuously(client, &**source)),
+        )
+        .await;
+        // to get ! type
+        values[0]
+    }
+
+    async fn update_source_continuously(&self, client: &Client, source: &dyn MetaSource) -> ! {
+        loop {
+            MetaSource::wait_for_update(source).await;
+            if let Ok(meta) = retry(
+                "update_source",
+                // very slow retry
+                FibonacciBackoff::default()
+                    .with_min_delay(Duration::from_secs(10))
+                    .with_max_delay(Duration::from_secs(10 * 60))
+                    .with_max_times(usize::MAX),
+                || source.fetch(client),
+            )
+            .await
+            {
+                let mut dbtx = client.db().begin_transaction().await;
+                Self::update_meta_values(&mut dbtx, source.source_id(), &meta).await;
+                // TODO: what should have happen in case of conflicts?
+                // should we ignore the conflict or auto commit
+                let _ = dbtx.commit_tx_result().await;
+            }
+        }
+    }
+
+    async fn update_meta_values(
+        dbtx: &mut DatabaseTransaction<'_, Committable>,
+        source_id: u64,
+        meta: &BTreeMap<MetaFieldName, MetaFieldValue>,
+    ) {
+        let stream = dbtx.find_by_prefix(&MetaFieldPrefix).await;
+
+        let to_remove = stream
+            .filter(|(key, _)| std::future::ready(key.source_id == source_id))
+            .map(|(key, _)| key)
+            .collect::<Vec<_>>()
+            .await;
+
+        for key in to_remove {
+            dbtx.remove_entry(&key).await;
+        }
+        for (name, value) in meta {
+            dbtx.insert_entry(
+                &MetaFieldKey {
+                    name: name.clone(),
+                    source_id,
+                },
+                value,
+            )
+            .await;
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+#[non_exhaustive]
+pub struct LegacyClientConfigMetaSource {}
+
+pub const LEGACY_CLIENT_CONFIG_PRIORITY: u64 = 0;
+pub const LEGACY_CLIENT_CONFIG_SOURCE_ID: u64 = 0;
+
+#[apply(async_trait_maybe_send!)]
+impl MetaSource for LegacyClientConfigMetaSource {
+    fn source_id(&self) -> u64 {
+        LEGACY_CLIENT_CONFIG_SOURCE_ID
+    }
+
+    async fn wait_for_update(&self) {
+        pending().await
+    }
+
+    async fn fetch(
+        &self,
+        client: &Client,
+    ) -> anyhow::Result<BTreeMap<MetaFieldName, MetaFieldValue>> {
+        Ok(client
+            .config()
+            .global
+            .meta
+            .iter()
+            .map(|(key, value)| {
+                (
+                    MetaFieldName(key.to_owned()),
+                    MetaFieldValue {
+                        value: value.clone(),
+                        revision: Some(0), // config is constant
+                        priority: LEGACY_CLIENT_CONFIG_PRIORITY,
```

Comment:
Giving each value a priority chosen by the source seems like over-engineering given that we have quite a clear precedence that is dependent on the source (2 legacy, one current). I don't think we'll ever have more than one non-legacy source.

Reply:
fixed